### ISSUE TITLE
Parsing error on dependent template name

### ISSUE THREAD

**Issue by JanValiska** [tree-sitter/tree-sitter-cpp#107]
Hi,

to reproduce my problem, just try parse this example:
```cpp
#include <stddef.h>

template<class T, class D>
struct S {
	static constexpr size_t I = T::template get<S>();
};
```

Result is parsed with error on 4th line ```ERROR [4, 41] - [4, 44]```:
```
preproc_include [0, 0] - [2, 0]
  path: system_lib_string [0, 9] - [0, 19]
template_declaration [2, 0] - [5, 2]
  parameters: template_parameter_list [2, 8] - [2, 26]
    type_parameter_declaration [2, 9] - [2, 16]
      type_identifier [2, 15] - [2, 16]
    type_parameter_declaration [2, 18] - [2, 25]
      type_identifier [2, 24] - [2, 25]
  struct_specifier [3, 0] - [5, 1]
    name: type_identifier [3, 7] - [3, 8]
    body: field_declaration_list [3, 9] - [5, 1]
      field_declaration [4, 1] - [4, 50]
        storage_class_specifier [4, 1] - [4, 7]
        type_qualifier [4, 8] - [4, 17]
        type: primitive_type [4, 18] - [4, 24]
        declarator: field_identifier [4, 25] - [4, 26]
        default_value: call_expression [4, 29] - [4, 49]
          function: template_function [4, 29] - [4, 47]
            name: scoped_identifier [4, 29] - [4, 40]
              namespace: namespace_identifier [4, 29] - [4, 30]
              name: identifier [4, 32] - [4, 40]
            ERROR [4, 41] - [4, 44]
              identifier [4, 41] - [4, 44]
            arguments: template_argument_list [4, 44] - [4, 47]
              type_descriptor [4, 45] - [4, 46]
                type: type_identifier [4, 45] - [4, 46]
          arguments: argument_list [4, 47] - [4, 49]
```

I'm using nightly neovim and latest tree-sitter plugin. Parser is updated to latest version too. Thank you.

**Comment by narpfel** [comment]
Two more examples:

Dependent member function templates (also allowed with `->`) ([grammar](https://eel.is/c++draft/expr.post.general))
```cpp
template<auto x>
auto foo(auto bar) {
  return bar.template get<x>();
}
```
which is parsed as
```
(translation_unit [0, 0] - [4, 0]
  (template_declaration [0, 0] - [3, 1]
    parameters: (template_parameter_list [0, 8] - [0, 16]
      (parameter_declaration [0, 9] - [0, 15]
        type: (auto [0, 9] - [0, 13])
        declarator: (identifier [0, 14] - [0, 15])))
    (function_definition [1, 0] - [3, 1]
      type: (auto [1, 0] - [1, 4])
      declarator: (function_declarator [1, 5] - [1, 18]
        declarator: (identifier [1, 5] - [1, 8])
        parameters: (parameter_list [1, 8] - [1, 18]
          (parameter_declaration [1, 9] - [1, 17]
            type: (auto [1, 9] - [1, 13])
            declarator: (identifier [1, 14] - [1, 17]))))
      body: (compound_statement [1, 19] - [3, 1]
        (return_statement [2, 2] - [2, 31]
          (call_expression [2, 9] - [2, 30]
            function: (field_expression [2, 9] - [2, 28]
              argument: (identifier [2, 9] - [2, 12])
              field: (template_method [2, 13] - [2, 28]
                name: (field_identifier [2, 13] - [2, 21])
                (ERROR [2, 22] - [2, 25]
                  (identifier [2, 22] - [2, 25]))
                arguments: (template_argument_list [2, 25] - [2, 28]
                  (type_descriptor [2, 26] - [2, 27]
                    type: (type_identifier [2, 26] - [2, 27])))))
            arguments: (argument_list [2, 28] - [2, 30])))))))
dependent_member_function.cxx	0 ms	(ERROR [2, 22] - [2, 25])
```
and dependent inner class templates ([grammar](https://eel.is/c++draft/expr.prim.id.qual))
```cpp
template<typename Foo>
auto f() {
    return typename Foo::template Bar<int>{42}.bar;
}
```
which is parsed as
```
(translation_unit [0, 0] - [4, 0]
  (template_declaration [0, 0] - [3, 1]
    parameters: (template_parameter_list [0, 8] - [0, 20]
      (type_parameter_declaration [0, 9] - [0, 19]
        (type_identifier [0, 18] - [0, 19])))
    (function_definition [1, 0] - [3, 1]
      type: (auto [1, 0] - [1, 4])
      declarator: (function_declarator [1, 5] - [1, 9]
        declarator: (identifier [1, 5] - [1, 6])
        parameters: (parameter_list [1, 6] - [1, 9]
          (parameter_declaration [1, 7] - [1, 8]
            type: (type_identifier [1, 7] - [1, 8]))))
      body: (compound_statement [1, 10] - [3, 1]
        (return_statement [2, 4] - [2, 19]
          (identifier [2, 11] - [2, 19]))
        (declaration [2, 20] - [2, 47]
          type: (scoped_type_identifier [2, 20] - [2, 31]
            namespace: (namespace_identifier [2, 20] - [2, 21])
            name: (type_identifier [2, 23] - [2, 31]))
          declarator: (init_declarator [2, 32] - [2, 44]
            declarator: (template_function [2, 32] - [2, 40]
              name: (identifier [2, 32] - [2, 35])
              arguments: (template_argument_list [2, 35] - [2, 40]
                (type_descriptor [2, 36] - [2, 39]
                  type: (primitive_type [2, 36] - [2, 39]))))
            value: (initializer_list [2, 40] - [2, 44]
              (number_literal [2, 41] - [2, 43])))
          (ERROR [2, 44] - [2, 46]
            (identifier [2, 45] - [2, 46])))))))
dependent_class_template.cxx	0 ms	(MISSING ";" [2, 19] - [2, 19])
```
In the second example, `typename` is required and `template` is optional but allowed.

**Comment by jdrouhard** [comment]
Fixed in #133
